Fix bsb parser bugs + add Jest

jchavarri opened the pull request:
Fixes #51.
Fixes #83.

Adds some basic testing infra using Jest + snapshots.

freebroccolo commented on package.json:
```diff
@@ -27,7 +27,15 @@
+  "jest": {
+    "moduleFileExtensions": ["ts", "tsx", "js"],
+    "transform": {
+      "^.+\\.(ts|tsx)$": "<rootDir>/preprocessor.js"
```
Can we move this file to `script/preprocessor.js`?

freebroccolo commented on preprocessor.js:
```diff
@@ -0,0 +1,13 @@
+    }
+    return src;
+  },
+};
```
Can we rewrite this as something like the following? I would prefer to use es6 and to avoid pulling in the copyright.

```javascript
import tsc from "typescript";
import tsConfig from "./tsconfig.json";

export function process(src, path) {
  if (/\.(tsx?)$/.test(path)) {
    return tsc.transpile(src, tsConfig.compilerOptions, path, []);
  } else {
    return src;
  }
}
```

jchavarri commented:
@freebroccolo I've added [`ts-jest`](https://github.com/kulshekhar/ts-jest) which includes the transform and sourcemap support, so I deleted `preprocessor.js` altogether.

> I haven't tried the Jest stuff but it seems like a good idea.

Yeah, Jest makes this kind of testing really easy. Ideally, we could add `bs-platform` as devDependency and call `bsb` straight away, but I thought it added too much complexity.

If we need to update the snapshots, there's `jest -u` so it makes things pretty fast (one has to be careful and always review the produced snapshot, in order to not store faulty ones).

> It would be kind of nice if we could split this into two separate PRs but it's not critical.

I thought about that, but I wanted to have Jest available for those two parser fixes which would require the 2nd PR with the fixes to depend on the 1st branch with Jest addition, so just went with one PR to avoid the overhead.

I've also been thinking it'd be nice to run these tests in TravisCI, but because all this `bsb` parser part is a bit in flux, for now i'll just keep running the tests manually before submitting a PR, unless you think otherwise.

freebroccolo commented:
> I thought about that, but I wanted to have Jest available for those two parser fixes which would require the 2nd PR with the fixes to depend on the 1st branch with Jest addition, so just went with one PR to avoid the overhead.

Ah, makes sense then. Thanks for the clarification.

> I've also been thinking it'd be nice to run these tests in TravisCI, but because all this bsb parser part is a bit in flux, for now i'll just keep running the tests manually before submitting a PR, unless you think otherwise.

Yeah, that's probably easier for now. I agree that we should focus on CI and automated tests soon though.
